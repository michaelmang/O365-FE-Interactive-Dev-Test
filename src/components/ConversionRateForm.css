.container {
  display: flex;
  flex-direction: row;
  justify-content: center;
}

.field {
  margin-right: 20px;
}

.field:last-child {
  margin-right: 0;
}
