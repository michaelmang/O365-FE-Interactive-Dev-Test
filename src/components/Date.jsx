import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import styles from './Date.module.scss';

const Date = ({ onDate, value }) => (
  <div className={classnames('field', styles.field)}>
    <div className="control">
      <input className="input" onChange={onDate} type="date" value={value} />
    </div>
  </div>
);

Date.propTypes = {
  onDate: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

export default Date;
