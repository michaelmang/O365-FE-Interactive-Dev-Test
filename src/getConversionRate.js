import axios from 'axios';

const getConversionRate = async ({ baseCurrency, date, targetCurrency }={}) => {
  if (!baseCurrency) {
    throw new Error('Must provide baseCurrency.');
  }

  if (!date) {
    throw new Error('Must provide date.');
  }

  if (!targetCurrency) {
    throw new Error('Must provide targetCurrency.');
  }

  let url = `https://api.exchangeratesapi.io/${date}?base=${baseCurrency}`;
  let resp = await axios.get(url);
  return getTargetCurrency(targetCurrency, resp);
};

const getTargetCurrency = (targetCurrency, { data }) => data.rates[targetCurrency];

export default getConversionRate;
