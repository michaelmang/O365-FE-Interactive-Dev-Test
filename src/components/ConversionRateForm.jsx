import PropTypes from 'prop-types';
import React from 'react';

import styles from './ConversionRateForm.module.scss';
import Date from './Date.jsx';
import Select from './Select.jsx';
import getConversionRate from '../getConversionRate.js';
import formatRate from '../helpers/formatRate.js';

export default class ConversionRateForm extends React.PureComponent {
  state = {
    baseCurrency: '',
    conversionRate: null,
    date: '',
    isLoading: false,
    targetCurrency: '',
  }

  handleApiException(ex) {
    this.setState({ isLoading: false });
    this.props.appendError(ex);
    console.error(ex);
    return;
  }

  handleBaseCurrency = ({ target }) => {
    let { value: baseCurrency } = target;
    this.setState({ baseCurrency });

    this.resetConversionRate();
  }

  handleDate = ({ target }) => {
    let { value: date } = target;
    this.setState({ date });

    this.resetConversionRate();
  }

  handleSubmit = async (event) => {
    event.preventDefault();

    this.setState({ isLoading: true });

    let conversionRate;
    try {
      this.validateSubmission();

      let { baseCurrency, date, targetCurrency } = this.state;
      conversionRate = await getConversionRate({
        baseCurrency,
        date,
        targetCurrency,
      });
    }
    catch (ex) {
      this.handleApiException(ex);
    }
    finally {
      this.setState({ isLoading: false });
      this.setState({ conversionRate });
    }
  }

  handleTargetCurrency = ({ target }) => {
    let { value: targetCurrency } = target;
    this.setState({ targetCurrency });

    this.resetConversionRate();
  }

  resetConversionRate() {
    this.setState({ conversionRate: null });
  }

  validateSubmission() {
    let { baseCurrency, date, targetCurrency } = this.state;

    if (!baseCurrency || hasDefaultOption(baseCurrency)) {
      throw new Error('Must provide a valid base currency.');
    }

    if (!targetCurrency || hasDefaultOption(targetCurrency)) {
      throw new Error('Must provide a valid target currency.');
    }

    if (!date) {
      throw new Error('Must provide a valid date.');
    }
  }

  render() {
    let { conversionRate, isLoading } = this.state;

    return (
      <div className={styles.container}>
        {isLoading ? this.renderLoader() : this.renderForm()}
        {(!isLoading && !!conversionRate) && this.renderConversionRate()}
      </div>
    );
  }

  renderConversionRate() {
    let {
      baseCurrency,
      conversionRate,
      date,
      targetCurrency,
    } = this.state;

    return (
      <div className={styles.container} data-test-id="conversion-rate-output">
        <p>1 {baseCurrency} =</p>
        <p className="has-text-dark has-text-weight-semibold is-size-4">
          {formatRate(conversionRate)} {targetCurrency}
        </p>
        <p className="has-text-grey">as of {date}</p>
      </div>
    );
  }

  renderForm() {
    let {
      baseCurrency,
      date,
      targetCurrency,
    } = this.state;

    return (
      <form className={styles.form} onSubmit={(e) => this.handleSubmit(e)}>
        <Select
          label="Pick Base Currency"
          onSelect={this.handleBaseCurrency}
          value={baseCurrency}
        />
        <Select
          label="Pick Target Currency"
          onSelect={this.handleTargetCurrency}
          value={targetCurrency}
        />
        <Date onDate={this.handleDate} value={date} />
        <input className="button is-primary" type="submit" value="Look Up" />
      </form>
    );
  }

  renderLoader() {
    return (
      <progress
        className="progress is-medium is-primary"
        max="100"
        style={{ width: '20rem' }}
      />
    );
  }
}

const hasDefaultOption = (val) => val === '-1';

ConversionRateForm.propTypes = {
  appendError: PropTypes.func.isRequired,
};
