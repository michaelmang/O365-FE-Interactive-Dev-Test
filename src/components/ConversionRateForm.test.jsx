import { shallow } from 'enzyme';
import React from 'react';

import ConversionRateForm from './ConversionRateForm.jsx';
import Date from './Date.jsx';
import getConversionRate from '../getConversionRate.js';
import renderSnapshot from '../testing/renderSnapshot.js';

jest.mock('./ConversionRateForm.module.scss', () => ({
  container: 'styles.container',
}));
jest.mock('./Date.jsx', () => 'Date');
jest.mock('./Select.jsx', () => 'Select');
jest.mock('../getConversionRate.js');

describe('ConversionRateForm', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    jest.resetAllMocks();
  });

  it('matches the saved snapshot', () => {
    let { root: actual } = extractComponents(renderComponent());

    expect(renderSnapshot(actual)).toMatchSnapshot();
  });

  describe('base currency select', () => {
    describe('when onSelect fires', () => {
      it('passes down value prop set to received event value', () => {
        let { baseCurrencySelect: subject, root } = extractComponents(renderComponent());

        subject.props().onSelect(createEvent('some base currency'));
        root.update();
        subject = extractComponents(root).baseCurrencySelect;

        expect(subject).toHaveProp(
          'value',
          'some base currency',
        );
      });

      it('resets conversionRate state', () => {
        let { baseCurrencySelect: subject, root } = extractComponents(renderComponent());
        root.setState({
          conversionRate: '123',
        });

        subject.props().onSelect(createEvent('some base currency'));
        root.update();

        expect(root).toHaveState(
          'conversionRate',
          null,
        );
      });
    });
  });

  describe('target currency select', () => {
    describe('when onSelect fires', () => {
      it('passes down value prop set to received event value', () => {
        let { targetCurrencySelect: subject, root } = extractComponents(renderComponent());

        subject.props().onSelect(createEvent('some target currency'));
        root.update();
        subject = extractComponents(root).targetCurrencySelect;

        expect(subject).toHaveProp(
          'value',
          'some target currency',
        );
      });

      it('resets conversionRate state', () => {
        let { targetCurrencySelect: subject, root } = extractComponents(renderComponent());
        root.setState({
          conversionRate: '123',
        });

        subject.props().onSelect(createEvent('some target currency'));
        root.update();

        expect(root).toHaveState(
          'conversionRate',
          null,
        );
      });
    });
  });

  describe('date select', () => {
    describe('when onDate fires', () => {
      it('passes down value prop set to received event value', () => {
        let { dateSelect: subject, root } = extractComponents(renderComponent());

        subject.props().onDate(createEvent('some date'));
        root.update();
        subject = extractComponents(root).dateSelect;

        expect(subject).toHaveProp(
          'value',
          'some date',
        );
      });

      it('resets conversionRate state', () => {
        let { dateSelect: subject, root } = extractComponents(renderComponent());
        root.setState({
          conversionRate: '123',
        });

        subject.props().onDate(createEvent('some date'));
        root.update();

        expect(root).toHaveState(
          'conversionRate',
          null,
        );
      });
    });
  });

  describe('on form submission', () => {
    it('calls event.preventDefault', () => {
      let preventDefault = jest.fn();
      let { form } = extractComponents(renderComponent());

      expect(preventDefault).not.toHaveBeenCalled();

      form.props().onSubmit({ preventDefault });

      expect(preventDefault).toHaveBeenCalled();
    });

    describe('submission validation', () => {
      describe('missing base currency', () => {
        it('calls appendError prop with associated exception', async () => {
          let appendError = jest.fn();
          let { form } = extractComponents(renderComponent({
            appendError,
          }));

          form.props().onSubmit({
            preventDefault: () => {},
          });

          let expected = new Error('Must provide a valid base currency.');
          expect(appendError).toHaveBeenCalledWith(expected);
        });

        it('also calls appendError with associated exception when default option selected', () => {
          let appendError = jest.fn();
          let { baseCurrencySelect, form } = extractComponents(renderComponent({
            appendError,
          }));

          baseCurrencySelect.props().onSelect(createDefaultOption());

          form.props().onSubmit({
            preventDefault: () => {},
          });

          let expected = new Error('Must provide a valid base currency.');
          expect(appendError).toHaveBeenCalledWith(expected);
        });

        it('renders form and no conversion rate output', async () => {
          let { form, root } = extractComponents(renderComponent());

          form.props().onSubmit({
            preventDefault: () => {},
          });
          root.update();

          expect(renderSnapshot(root)).toMatchSnapshot();
        });
      });

      describe('missing target currency', () => {
        it('calls appendError prop with associated exception', async () => {
          let appendError = jest.fn();
          let {
            baseCurrencySelect,
            form,
          } = extractComponents(renderComponent({
            appendError,
          }));

          baseCurrencySelect.props().onSelect(createEvent('some base currency'));

          form.props().onSubmit({
            preventDefault: () => {},
          });

          let expected = new Error('Must provide a valid target currency.');
          expect(appendError).toHaveBeenCalledWith(expected);
        });


        it('also calls appendError with associated exception when default option selected', () => {
          let appendError = jest.fn();
          let {
            baseCurrencySelect,
            form,
            targetCurrencySelect,
          } = extractComponents(renderComponent({
            appendError,
          }));

          baseCurrencySelect.props().onSelect(createEvent('some base currency'));
          targetCurrencySelect.props().onSelect(createDefaultOption());

          form.props().onSubmit({
            preventDefault: () => {},
          });

          let expected = new Error('Must provide a valid target currency.');
          expect(appendError).toHaveBeenCalledWith(expected);
        });

        it('renders form and no conversion rate output', async () => {
          let {
            baseCurrencySelect,
            form,
            root,
          } = extractComponents(renderComponent());

          baseCurrencySelect.props().onSelect(createEvent('some base currency'));

          form.props().onSubmit({
            preventDefault: () => {},
          });
          root.update();

          expect(renderSnapshot(root)).toMatchSnapshot();
        });
      });

      describe('missing date', () => {
        it('calls appendError prop with associated exception', async () => {
          let appendError = jest.fn();
          let {
            baseCurrencySelect,
            form,
            targetCurrencySelect,
          } = extractComponents(renderComponent({
            appendError,
          }));

          baseCurrencySelect.props().onSelect(createEvent('some base currency'));
          targetCurrencySelect.props().onSelect(createEvent('some target currency'));

          form.props().onSubmit({
            preventDefault: () => {},
          });

          let expected = new Error('Must provide a valid date.');
          expect(appendError).toHaveBeenCalledWith(expected);
        });

        it('renders form and no conversion rate output', async () => {
          let {
            baseCurrencySelect,
            form,
            root,
            targetCurrencySelect,
          } = extractComponents(renderComponent());

          baseCurrencySelect.props().onSelect(createEvent('some base currency'));
          targetCurrencySelect.props().onSelect(createEvent('some target currency'));

          form.props().onSubmit({
            preventDefault: () => {},
          });
          root.update();

          expect(renderSnapshot(root)).toMatchSnapshot();
        });
      });
    });

    it('calls getConversionRate with baseCurrencySelect, date, and targetCurrency state', () => {
      let {
        baseCurrencySelect,
        dateSelect,
        form,
        targetCurrencySelect,
      } = extractComponents(renderComponent());

      baseCurrencySelect.props().onSelect(createEvent('some base currency'));
      targetCurrencySelect.props().onSelect(createEvent('some target currency'));
      dateSelect.props().onDate(createEvent('some date'));

      expect(getConversionRate).not.toHaveBeenCalled();

      form.props().onSubmit({
        preventDefault: () => {},
      });

      expect(getConversionRate).toHaveBeenCalledWith({
        baseCurrency: 'some base currency',
        date: 'some date',
        targetCurrency: 'some target currency',
      });
    });

    describe('when getConversionRate is pending', () => {
      it('only renders progress bar', () => {
        let {
          baseCurrencySelect,
          dateSelect,
          form,
          root,
          targetCurrencySelect,
        } = extractComponents(renderComponent());

        baseCurrencySelect.props().onSelect(createEvent('some base currency'));
        targetCurrencySelect.props().onSelect(createEvent('some target currency'));
        dateSelect.props().onDate(createEvent('some date'));

        form.props().onSubmit({
          preventDefault: () => {},
        });
        root.update();

        expect(renderSnapshot(root)).toMatchSnapshot();
      });
    });

    describe('when getConversionRate resolves successfully', () => {
      beforeEach(() => {
        getConversionRate.mockResolvedValue('123.4567');
      });

      it('renders form and conversion rate output', async () => {
        let actual = await submitForm();

        expect(renderSnapshot(actual)).toMatchSnapshot();
      });
    });

    describe('when getConversionRate throws an exception', () => {
      let someError;

      beforeEach(() => {
        someError = new Error('kaboom');
        getConversionRate.mockRejectedValue(someError);
      });

      it('calls appendError prop with raised exception', async () => {
        let appendError = jest.fn();

        await submitForm({ appendError });

        expect(appendError).toHaveBeenCalledWith(someError);
      });

      it('renders form and no conversion rate output', async () => {
        let appendError = jest.fn();

        let actual = await submitForm({ appendError });

        expect(renderSnapshot(actual)).toMatchSnapshot();
      });
    });
  });

  const createDefaultOption = () => createEvent('-1');

  const createEvent = (value) => ({ target: { value } });

  const extractComponents = (root) => ({
    baseCurrencySelect: root.find('[label="Pick Base Currency"]'),
    dateSelect: root.find(Date),
    form: root.find('form'),
    root,
    targetCurrencySelect: root.find('[label="Pick Target Currency"]'),
  });

  const renderComponent = (props) => (
    shallow(
      <ConversionRateForm
        appendError={() => {}}
        {...props}
      />
    )
  );

  const submitForm = async (props) => {
    let {
      baseCurrencySelect,
      dateSelect,
      form,
      root,
      targetCurrencySelect,
    } = extractComponents(renderComponent(props));

    baseCurrencySelect.props().onSelect(createEvent('some base currency'));
    targetCurrencySelect.props().onSelect(createEvent('some target currency'));
    dateSelect.props().onDate(createEvent('some date'));

    await form.props().onSubmit({
      preventDefault: () => {},
    });
    root.update();

    return root;
  };
});
