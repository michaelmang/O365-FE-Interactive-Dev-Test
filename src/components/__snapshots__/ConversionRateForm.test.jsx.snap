// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionRateForm matches the saved snapshot 1`] = `
<div
  className="styles.container"
>
  <form
    onSubmit={[Function]}
  >
    <Select
      label="Pick Base Currency"
      onSelect={[Function]}
      value=""
    />
    <Select
      label="Pick Target Currency"
      onSelect={[Function]}
      value=""
    />
    <Date
      onDate={[Function]}
      value=""
    />
    <input
      className="button is-primary"
      type="submit"
      value="Look Up"
    />
  </form>
</div>
`;

exports[`ConversionRateForm on form submission submission validation missing base currency renders form and no conversion rate output 1`] = `
<div
  className="styles.container"
>
  <form
    onSubmit={[Function]}
  >
    <Select
      label="Pick Base Currency"
      onSelect={[Function]}
      value=""
    />
    <Select
      label="Pick Target Currency"
      onSelect={[Function]}
      value=""
    />
    <Date
      onDate={[Function]}
      value=""
    />
    <input
      className="button is-primary"
      type="submit"
      value="Look Up"
    />
  </form>
</div>
`;

exports[`ConversionRateForm on form submission submission validation missing date renders form and no conversion rate output 1`] = `
<div
  className="styles.container"
>
  <form
    onSubmit={[Function]}
  >
    <Select
      label="Pick Base Currency"
      onSelect={[Function]}
      value="some base currency"
    />
    <Select
      label="Pick Target Currency"
      onSelect={[Function]}
      value="some target currency"
    />
    <Date
      onDate={[Function]}
      value=""
    />
    <input
      className="button is-primary"
      type="submit"
      value="Look Up"
    />
  </form>
</div>
`;

exports[`ConversionRateForm on form submission submission validation missing target currency renders form and no conversion rate output 1`] = `
<div
  className="styles.container"
>
  <form
    onSubmit={[Function]}
  >
    <Select
      label="Pick Base Currency"
      onSelect={[Function]}
      value="some base currency"
    />
    <Select
      label="Pick Target Currency"
      onSelect={[Function]}
      value=""
    />
    <Date
      onDate={[Function]}
      value=""
    />
    <input
      className="button is-primary"
      type="submit"
      value="Look Up"
    />
  </form>
</div>
`;

exports[`ConversionRateForm on form submission when getConversionRate is pending only renders progress bar 1`] = `
<div
  className="styles.container"
>
  <progress
    className="progress is-medium is-primary"
    max="100"
    style={
      Object {
        "width": "20rem",
      }
    }
  />
</div>
`;

exports[`ConversionRateForm on form submission when getConversionRate resolves successfully renders form and conversion rate output 1`] = `
<div
  className="styles.container"
>
  <form
    onSubmit={[Function]}
  >
    <Select
      label="Pick Base Currency"
      onSelect={[Function]}
      value="some base currency"
    />
    <Select
      label="Pick Target Currency"
      onSelect={[Function]}
      value="some target currency"
    />
    <Date
      onDate={[Function]}
      value="some date"
    />
    <input
      className="button is-primary"
      type="submit"
      value="Look Up"
    />
  </form>
  <div
    className="styles.container"
    data-test-id="conversion-rate-output"
  >
    <p>
      1 
      some base currency
       =
    </p>
    <p
      className="has-text-dark has-text-weight-semibold is-size-4"
    >
      123.457
       
      some target currency
    </p>
    <p
      className="has-text-grey"
    >
      as of 
      some date
    </p>
  </div>
</div>
`;

exports[`ConversionRateForm on form submission when getConversionRate throws an exception renders form and no conversion rate output 1`] = `
<div
  className="styles.container"
>
  <form
    onSubmit={[Function]}
  >
    <Select
      label="Pick Base Currency"
      onSelect={[Function]}
      value="some base currency"
    />
    <Select
      label="Pick Target Currency"
      onSelect={[Function]}
      value="some target currency"
    />
    <Date
      onDate={[Function]}
      value="some date"
    />
    <input
      className="button is-primary"
      type="submit"
      value="Look Up"
    />
  </form>
</div>
`;
